### DISCUSSION TITLE
Is it possible for a controller to own a dynamic type?

### DISCUSSION THREAD

**Discussion by empath-nirvana** [kube-rs/kube#922] [General]
i have a controller that updates a third party resource that has a fairly complex spec that I don't really feel like fully implementing, so I did it with a dynamic type, but I'm trying to figure out how to make it own the type and running into this error:

`type mismatch resolving `<DynamicObject as kube::Resource>::DynamicType == ()`

Do I have to implement every 3rd party struct that I want to own with a controller or is there a way to own a dynamically typed object?

**Comment by clux** [comment]
There's a few options available to you noted in the controller-guide under [dynamic typing](https://kube.rs/controllers/object/#dynamic-typing). Basically, either use `DynamicObject` or a partially typed `Object` and pass the `ApiResource` for that to to `Controller::owns_with`.

**Comment by nightkr** [comment]
So there are really two questions involved here.

1. Do I really have to implement every field when interacting with foreign CRDs?

No! Unknown fields will be ignored, you only need to add the fields you actually care about reading. Just make sure to not apply the generated CRD definition in this case.

2. Can I use dynamically typed `Resource` types with `Controller`?

Yes! You just need to use the `_with` (`new_with`, `owns_with`, `watches_with`) variants and provide the resource metadata yourself.
